internal/engine/buildcontroller_test.go

```diff
@@ -1047,3 +1048,303 @@ func TestLogsLongResourceName(t *testing.T) {
+	f.editFileAndAssertManifestNotBuilding("manB", "b/main.go")
+
+	// resync the two counts, manB build will start as expected
+	indexB := f.runAndWaitForManifestBuilding("manB", func() {
+		st := f.store.LockMutableStateForTesting()
+		st.StartedBuildCount++
+		f.store.UnlockMutableState()
+	})
+
+	// complete all builds (can't guarantee order)
+	f.completeAndCheckBuildsForManifests([]int{indexA, indexB}, manifests)
+
+	err := f.Stop()
+	assert.NoError(t, err)
+	f.assertAllBuildsConsumed()
+}
+
+func TestErrorHandlingWithMultipleBuilds(t *testing.T) {
+	f := newTestFixture(t)
+	defer f.TearDown()
+	f.b.completeBuildsManually = true
+	f.setMaxBuildSlots(2)
+
+	errA := fmt.Errorf("errA")
+	errB := fmt.Errorf("errB")
+
+	manA := f.newDockerBuildManifestWithBuildPath("manA", f.JoinPath("a"))
+	manB := f.newDockerBuildManifestWithBuildPath("manB", f.JoinPath("b"))
+	manC := f.newDockerBuildManifestWithBuildPath("manC", f.JoinPath("c"))
+	f.StartAndWaitForInitialBuilds([]model.Manifest{manA, manB, manC})
+
+	// start builds for all manifests (we only have 2 build slots)
+	f.SetNextBuildFailure(errA)
+	indexA := f.editFileAndWaitForManifestBuilding("manA", "a/main.go")
+	f.SetNextBuildFailure(errB)
+	indexB := f.editFileAndWaitForManifestBuilding("manB", "b/main.go")
+	f.editFileAndAssertManifestNotBuilding("manC", "c/main.go")
+
+	// Complete one build -- 'manC' should start building, even though the manA build ended with an error
+	indexC := f.runAndWaitForManifestBuilding("manC", func() {
+		f.b.completeBuild(indexA)
+		call := f.nextCall("expect manA build complete")
+		f.assertCallIsForManifestAndFiles(call, manA, "a/main.go")
+		f.WaitUntilManifestState("last manA build reflects expected error", "manA", func(ms store.ManifestState) bool {
+			return ms.LastBuild().Error == errA
+		})
+	})
+
+	// complete the rest
+	f.completeAndCheckBuildsForManifests([]int{indexB, indexC}, []model.Manifest{manB, manC})
+	f.WaitUntilManifestState("last manB build reflects expected error", "manB", func(ms store.ManifestState) bool {
+		return ms.LastBuild().Error == errB
+	})
+	time.Sleep(10 * time.Millisecond)
+	f.WaitUntilManifestState("last manC build has no error", "manC", func(ms store.ManifestState) bool {
+		return ms.LastBuild().Error == nil
+	})
+
+	err := f.Stop()
+	assert.NoError(t, err)
+	f.assertAllBuildsConsumed()
+}
+
+func (f *testFixture) waitUntilManifestBuilding(name model.ManifestName) {
+	msg := fmt.Sprintf("manifest %q is building", name)
+	f.WaitUntilManifestState(msg, name, func(ms store.ManifestState) bool {
+		return ms.IsBuilding()
+	})
+
+	f.withState(func(st store.EngineState) {
+		_, ok := st.CurrentlyBuilding[name]
+		require.True(f.t, ok, "expected EngineState to reflect that %q is currently building", name)
+	})
+}
+
+func (f *testFixture) waitUntilBuildCountAtLeast(n int) {
+	ctx, cancel := context.WithTimeout(f.ctx, time.Millisecond*200)
+	defer cancel()
+	for {
+		if f.b.buildCount >= n {
```

Comment:
synchronizing on which `buildCount`?

so here's an example of where I'm using this func:
```
func (f *testFixture) editFileAndWaitForManifestBuilding(name model.ManifestName, path string) (buildIndex int) {
        startBuildCount := f.b.buildCount
        f.fsWatcher.events <- watch.NewFileEvent(f.JoinPath(path))
        f.waitUntilManifestBuilding(name)
        f.waitUntilBuildCountAtLeast(startBuildCount + 1)
        return startBuildCount + 1
}
```
`f.waitUntilManifestBuilding(name)` checks that the manifest is building _according to the engine state_, i.e. that we've processed the `BuildStartedAction`. The point of `waitUntilBuildCountAtLeast` (which I should probably rename) is to make sure that this build has hit the BaD (and that the index that we return definitely corresponds to this build).

Without this check, I ran into problems with something like:

```
f.setMaxBuildSlots(3)
indexA := f.editFileAndWaitForManifestBuilding("manA", "a/main.go")
indexB := f.editFileAndWaitForManifestBuilding("manB", "b/main.go")
indexC := f.editFileAndWaitForManifestBuilding("manC",  "c/main.go")
f.b.completeBuild(indexB)
```
because of the order in which these manifests actually hit the buildAndDeployer, the indexes we associated with them were NOT a simple matter of `f.b.buildCount++` for each; e.g. here's a test result where resolving the build at `indexB` didn't actually correspond to a build of manifestB:
```

--- FAIL: TestCurrentlyBuildingMayExceedMaxBuildCount (0.01s)
    buildcontroller_test.go:1147: 
                Error Trace:        buildcontroller_test.go:1147
                Error:              Not equal: 
                                    expected: model.TargetID{Type:"image", Name:"manb"}
                                    actual  : model.TargetID{Type:"image", Name:"mana"}
                                    
                                    Diff:
                                    --- Expected
                                    +++ Actual
                                    @@ -1,2 +1,2 @@
                                    -(model.TargetID) image:manb
                                    +(model.TargetID) image:mana
                                     
                Test:               TestCurrentlyBuildingMayExceedMaxBuildCount
```
BUT I would love a better way of doing this!

Reply:
you're reading f.b.buildCount, so you need to f.b.mu.Lock() the mutex.

tho this whole discussion makes me wonder if the real problem is that we've structured the test harness wrong. Using indices as identifiers means you have to do a ton of work to enforce ordering. But the whole point of this code is that things should happen in parallel, and we're not actually exercising the parallelism.

what if we changed registerBuild/completeBuild to use target ids instead of indices? would that simplify things?